### PR TITLE
refactor v4 client to use httpx under-the-hood

### PR DESCRIPTION
**Pull request by tsmith023**

### CODE REVIEW

File: weaviate/config.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 @dataclass
 class ConnectionConfig:
     session_pool_connections: int = 20
-    session_pool_maxsize: int = 20
+    session_pool_maxsize: int = 100
```
Comment (dirkkul) [inline weaviate/config.py:10]: why this change?

File: weaviate/connect/auth.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            session = self._get_client_auth_bearer_token(self._credentials)
+        elif isinstance(self._credentials, AuthClientCredentials):
+            session = self._get_client_client_credential(self._credentials)
+        else:
+            assert isinstance(self._credentials, AuthClientPassword)
+            session = self._get_client_user_pw(self._credentials)
+
+        return session
+
+    async def get_async_auth_client(self) -> AsyncOAuth2Client:
```
Comment (dirkkul) [inline weaviate/connect/auth.py:86]: this is currently not used, right? There can always be only one oidc-auth connection

File: weaviate/connect/auth.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/connect/auth.py:1]: I will remove `auth.py` for now and only use `authentication.py`, but we will need to use an async oidc-auth class when extending to async for both so that it's non-blocking. We can revisit that later

File: weaviate/connect/auth.py (hunk 2 above)
Comment (tsmith023) [reply weaviate/connect/auth.py:86]: Yes, will remove this file as above

File: weaviate/client.py [hunk 4]
```diff
@@ -457,3 +451,7 @@ def timeout_config(self, timeout_config: TIMEOUT_TYPE) -> None:
         """
 
         self._connection.timeout_config = _get_valid_timeout_config(timeout_config)
+
+    def __del__(self) -> None:
+        # in case an exception happens before definition of the client
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/client.py:456]: The close method only closes the connections as its duty. Whereas, here, the client attempts to close any underlying connections if it gets deleted prematurely

File: weaviate/config.py (hunk 1 above)
Comment (tsmith023) [reply weaviate/config.py:10]: Only that these are the defaults used by `httpx` so I changed our config object to reflect that. The `session_pool_maxsize` variable controls the total number of allowed connections while `session_pool_connections` controls the number of keepalive connections. I reckon `20` for the max allowed is quite small considering

File: weaviate/connect/v4.py [hunk 5]
```diff
@@ -336,10 +371,6 @@ def _get_request_header(self) -> dict:
         """
         return self._headers
```
Comment (dirkkul) [inline weaviate/connect/v4.py:372]: delete this one, too?

**Review by dirkkul — APPROVED** [review]

File: weaviate/connect/v4.py (hunk 5 above)
Comment (tsmith023) [reply weaviate/connect/v4.py:372]: I left it as an abstraction in case we need to add more but it doesn't do much and if we do it's not an issue

**Review by dirkkul — APPROVED** [review]
